core/src/main/java/org/apache/iceberg/data/avro/GenericReaders.java

```diff
@@ -133,4 +143,40 @@ protected void set(Record struct, int pos, Object value) {
       struct.set(pos, value);
     }
   }
+
+  private static class InternalRecordReader<R extends Record> extends ValueReaders.StructReader<R> {
+    private final Class<R> recordClass;
+    private final DynConstructors.Ctor<R> ctor;
+    private final Schema schema;
+
+    InternalRecordReader(List<ValueReader<?>> readers, Class<R> recordClass, Schema schema,
+                         Map<Integer, ?> idToConstant) {
```

Comment:
I see the need for a reader that produces records of some specific class. I don't understand why that needs to accept an Avro schema, though. The only requirement should be that the class implements StructLike and accepts an Iceberg schema to set its projection.

Reply:
Yes, the `Record` implements the `StructLike`, so we need the Schema for setter and getter in case of projection.